src/confirm.rs

```diff
@@ -34,279 +27,93 @@ where
-        loop {
-            let next_state = match self.state {
-                SendTransactionWithConfirmationState::SendTransaction(ref mut future) => {
-                    let hash = ready!(future.poll_unpin(ctx))?;
-                    if self.confirmations > 0 {
-                        let confirmation_check =
-                            TransactionReceiptBlockNumberCheck::new(Eth::new(self.transport.clone()), hash);
-                        let eth = Eth::new(self.transport.clone());
-                        let eth_filter = EthFilter::new(self.transport.clone());
-                        let wait = wait_for_confirmations(
-                            eth,
-                            eth_filter,
-                            self.poll_interval,
-                            self.confirmations,
-                            confirmation_check,
-                        );
-                        SendTransactionWithConfirmationState::WaitForConfirmations(hash, wait)
-                    } else {
-                        let receipt_future = Eth::new(&self.transport).transaction_receipt(hash);
-                        SendTransactionWithConfirmationState::GetTransactionReceipt(receipt_future)
-                    }
-                }
-                SendTransactionWithConfirmationState::WaitForConfirmations(hash, ref mut future) => {
-                    let _confirmed = ready!(future.poll_unpin(ctx))?;
-                    let receipt_future = Eth::new(&self.transport).transaction_receipt(hash);
-                    SendTransactionWithConfirmationState::GetTransactionReceipt(receipt_future)
-                }
-                SendTransactionWithConfirmationState::GetTransactionReceipt(ref mut future) => {
-                    let receipt = ready!(future.poll_unpin(ctx))?
-                        .expect("receipt can't be null after wait for confirmations; qed");
-                    return Poll::Ready(Ok(receipt));
-                }
-            };
-            self.state = next_state;
-        }
+) -> error::Result<TransactionReceipt> {
+    let eth = Eth::new(transport.clone());
+    if confirmations > 0 {
+        let confirmation_check = || transaction_receipt_block_number_check(&eth, hash);
+        let eth_filter = EthFilter::new(transport.clone());
+        let eth = eth.clone();
+        wait_for_confirmations(eth, eth_filter, poll_interval, confirmations, confirmation_check).await?;
     }
+    // TODO: We should remove this `expect`. No matter what happens inside the node, this shouldn't be a panic.
+    // Should convert this into an some error. InternalError probably makes sense but then we cannot provide the error
+    // message. Maybe make new error for this? Or allow InternalError to carry a message.
+    let receipt = eth
+        .transaction_receipt(hash)
+        .await?
+        .expect("receipt can't be null after wait for confirmations; qed");
```

Comment:
This bug already existed so this PR doesn't make it worse.

Reply:
Maybe let's put the responsibility to `unwrap` on the user? I.e. returning `Option<TransactionReceipt>` here?

Reply:
Like above the issue was already there and I don't want to fix it in this PR. Added issue link.